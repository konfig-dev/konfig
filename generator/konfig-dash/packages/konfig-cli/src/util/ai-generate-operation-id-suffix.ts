import { Operation, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOperationIdSuffix({
  prefix,
  path,
  method,
  operation,
  operations,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  operations: OperationObject[]
}): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })

  const existingIds = operations.map((o) => o.operation.operationId)
  const existingIdsStr = existingIds.join(', ').slice(0, -2)
  const systemPrompt = `You are a generator of operation IDs for an OpenAPI specification. You generate operation IDs in json format: "{ prefix: ... , suffix: ...}"`
  const prompt = generatePrompt({
    prefix,
    path,
    method,
    operation,
    existingIdsStr,
  })

  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.suffix === undefined || responseJson.suffix === null)
    throw Error('OpenAI returned an invalid json: it does not contain a suffix')

  // Post-process to ensure that rules are followed
  let suffix = responseJson.suffix.replace('_', '')
  suffix = suffix.charAt(0).toLowerCase() + suffix.slice(1)
  suffix = suffix.replace('retrieve', 'get')
  suffix = suffix.replace('fetch', 'get')
  console.log(`AI-generated operation id: ${prefix}${suffix}`)
  return suffix
}

function generatePrompt({
  prefix,
  path,
  method,
  operation,
  existingIdsStr,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  existingIdsStr: string
}): string {
  return `Generate an operation ID for this operation in my OpenAPI specification:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}
Description: ${operation.description}

The operation ID is in the format "prefix_suffix". The prefix is: ${prefix}. Generate the suffix and respond in json format.
The operation ID MUST follow ALL of these rules:

1. The suffix must be at least 3 characters long.
2. The suffix must be in camelCase and CANNOT include any underscores.
3. The suffix must not redundantly re-use words from the prefix. (For example: Store_get is preferred over Store_getStore.)
4. If the suffix contains both a verb and a noun, the noun should generally follow the verb. (For example, Store_placeOrder is better than Store_orderPlace).
5. Prefer simple verbs like "get", "create", "update", "delete", and "list" when possible.

Existing operation ids for this OpenAPI specification are as follows: ${existingIdsStr}.
The operation ID MUST be unique within this list of existing operation IDs. You can also use them as a reference to generate this operation id.`
}
