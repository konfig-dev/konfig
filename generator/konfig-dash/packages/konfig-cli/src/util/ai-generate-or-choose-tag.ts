import { Operation, TagObject, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOrChooseTag(
  operation: Operation,
  path: string,
  method: string,
  tags: TagObject[]
): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })
  const tagNames = tags.map((t) => t.name)

  // First, try and choose a tag programmatically without using OpenAI
  const tag = programaticallyChooseTag(operation, path, tagNames)
  if (tag !== undefined) return tag
  const systemPrompt = `You are a generator of tags for an OpenAPI specification. You generate tags in json format: "{ tag: ... }"`
  const prompt = generatePrompt(path, method, operation, tagNames)
  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.tag === undefined || responseJson.tag === null)
    throw Error('OpenAI returned an invalid json: it does not contain a tag')

  tags.push({ name: responseJson.tag })
  return responseJson.tag
}

function programaticallyChooseTag(
  operation: Operation,
  path: string,
  tagNames: string[]
): string | undefined {
  for (const tag of tagNames) {
    if (path.toLowerCase().includes(tag.toLowerCase())) return tag
    if (
      operation.summary &&
      operation.summary.toLowerCase().includes(tag.toLowerCase())
    )
      return tag
  }
  return undefined
}

function generatePrompt(
  path: string,
  method: string,
  operation: Operation,
  tagNames: string[]
): string {
  const tagNamesStr = tagNames.join(', ').slice(0, -2)
  return `The following operation from my OpenAPI specification is missing a tag:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}

Based on this information, choose the best tag from this list of existing tags on the OpenApi Specification: ${tagNamesStr}.
IF AND ONLY IF none of these existing tags are suitable, then generate a new tag. The new tag MUST be a single word, MUST be a noun, and MUST be singular. It MUST NOT be a combination of multiple words.

In general, a good tag name is related to the type of data being operated on (such as "Student"), or the type of operation being performed (such as "Authentication").

Respond with just the tag in json format: "{ tag: ... }" with no other information.`
}
