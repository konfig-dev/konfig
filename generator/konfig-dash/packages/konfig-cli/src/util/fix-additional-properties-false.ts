import { Spec, recurseObject } from 'konfig-lib'

/**
 * "additionalProperties: false" causes issue when mocking an API.  The
 * following type of errors will appear for no reason: "âœ–  error     Request
 * body must NOT have additional properties; found 'userId'" but userId is
 * actually a property in the schema so the error is just wrong.  To get around
 * this I added this rule to remove "additionalProperties: false" from the
 * schema.
 */
export function fixAdditionalPropertiesFalse({ spec }: { spec: Spec }): number {
  let numberOfAdditionalPropertiesFixed = 0
  recurseObject(spec.spec, ({ value: schema }) => {
    if (typeof schema !== 'object') return
    if (schema === null) return
    if (schema.type !== 'object') return
    if (schema.additionalProperties === undefined) return
    if (schema.additionalProperties === true) return
    delete schema.additionalProperties
    numberOfAdditionalPropertiesFixed++
  })
  return numberOfAdditionalPropertiesFixed
}
