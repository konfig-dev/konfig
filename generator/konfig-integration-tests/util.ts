import { execa } from "execa";
import * as path from "path";
import yaml from "js-yaml";
import { expect } from "vitest";
import * as fs from "fs";
import {
  KONFIG_YAML_NAME,
  KonfigYaml,
  KonfigYamlAdditionalGeneratorConfig,
  KonfigYamlGeneratorConfig,
  KonfigYamlType,
} from "konfig-lib";
import waitOn from "wait-on";

const USE_RUN =
  process.env.GITHUB_ACTIONS === "true" ||
  (process.env.RUN && process.env.RUN.trim() !== "");

console.log("USE_RUN", USE_RUN);

// relative path from this file to: ""../konfig-dash/packages/konfig-cli/bin/dev""
// use nodejs __dirname to get the current directory of this file
const KONFIG_CLI_PATH = path.join(
  __dirname,
  `../konfig-dash/packages/konfig-cli/bin/${USE_RUN ? "run" : "dev"}`
);

/**
 * parse 'python-pydantic-empty-response' from 'tests/python-pydantic-empty-response.test.ts > python-pydantic-empty-response'
 */
function getCurrentTestName(): string {
  const currentTestName = expect.getState().currentTestName;
  const parsed = currentTestName?.split(" > ")[1];
  if (!parsed) {
    throw new Error(`Unable to parse current test name: ${currentTestName}`);
  }
  return parsed;
}

export async function e2e(
  mockServerPort: number,
  customAssertions?: () => void
) {
  if (!getCurrentTestName()) {
    throw new Error("Unable to get current test name");
  }
  const sdkDir = path.join(__dirname, "sdks", getCurrentTestName());

  const apiUrl = process.env.KONFIG_API_URL ?? "http://127.0.0.1:8911";

  // wait until KONFIG_API_URL/healthz is available
  await waitOn({
    resources: [`${apiUrl}/healthz`],
    timeout: 100000,
  });

  const env = {
    // This ensure oclif doesn't bug out
    // See: https://github.com/oclif/oclif/issues/1161#issuecomment-1661372245
    NODE_ENV: "",
  };

  // run "konfig generate -d -I" inside the path
  await execa(KONFIG_CLI_PATH, ["generate", "-d", "-I"], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // run "konfig test" inside the path
  await execa(KONFIG_CLI_PATH, ["test", "-p", mockServerPort.toString()], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // validate top-level README.md
  const konfigYamlPath = path.join(sdkDir, KONFIG_YAML_NAME);
  if (!fs.existsSync(konfigYamlPath))
    throw new Error(
      `Make sure you are in same directory as '${KONFIG_YAML_NAME}'`
    );
  const konfigYamlRaw = fs.readFileSync(konfigYamlPath, "utf-8");
  const loadedKonfigYaml: KonfigYamlType = yaml.load(
    konfigYamlRaw
  ) as KonfigYamlType;
  const parsedKonfigYaml = KonfigYaml.parse(loadedKonfigYaml);

  const generators: (
    | KonfigYamlGeneratorConfig
    | KonfigYamlAdditionalGeneratorConfig
  )[] = [
    ...Object.values(parsedKonfigYaml.generators),
    ...(parsedKonfigYaml.additionalGenerators !== undefined
      ? Object.values(parsedKonfigYaml.additionalGenerators)
      : []),
  ];
  for (const generator of generators) {
    expect(
      normalizeDocumentation(
        fs.readFileSync(
          path.join(sdkDir, generator.outputDirectory, "README.md"),
          "utf-8"
        )
      )
    ).toMatchSnapshot();
  }

  if (customAssertions) {
    customAssertions();
  }
}

// Removes the [GitHub last commit] line from the README
function normalizeDocumentation(readme: string) {
  // matches [GitHub last commit] to the end of the line
  const pattern = /\[!\[GitHub last commit\].*$\n?/gm;
  return readme.replace(pattern, "");
}
