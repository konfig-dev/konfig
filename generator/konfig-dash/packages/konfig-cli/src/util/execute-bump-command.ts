import { CliUx } from '@oclif/core'
import { getPackageVersion, setPackageVersion, KonfigYamlType } from 'konfig-lib'
import { generateKonfigYamlString } from '../util/generate-konfig-yaml-string'
import { incrementVersionWithPreservedTag } from '../util/increment-version-with-preserved-tag'
import semver from 'semver'
import * as fs from 'fs-extra'

export async function executeBumpCommand({
  filter,
  releaseType,
  loadedKonfigYaml,
  konfigYamlPath,
}: {
  filter: string[] | null
  releaseType: semver.ReleaseType
  loadedKonfigYaml: KonfigYamlType
  konfigYamlPath: string
}): Promise<void> {

  if (filter !== null)
    CliUx.ux.info(`Performing ${releaseType} bump of ${filter.join(', ')}`)
  else CliUx.ux.info(`Performing "${releaseType}" bump`)

  for (const [generatorName, generatorConfig] of [
    ...Object.entries(loadedKonfigYaml.generators),
    ...Object.entries(loadedKonfigYaml.additionalGenerators ?? []),
  ]) {
    if (filter !== null && !filter.includes(generatorName)) continue
    if ('disabled' in generatorConfig && generatorConfig.disabled) continue
    const currentVersion = getPackageVersion({ generatorConfig })
    const newVersion = incrementVersionWithPreservedTag(currentVersion, releaseType)
    CliUx.ux.info(`Bumping ${generatorName} SDK from ${currentVersion} --> ${newVersion}`)
    if (newVersion === null)
      CliUx.ux.error(`Could not compute new version for ${generatorName} SDK currently at ${currentVersion}`)
    setPackageVersion({ generatorConfig, version: newVersion })
  }
  CliUx.ux.action.start('Writing new version to konfig.yaml')
  fs.writeFileSync(konfigYamlPath, generateKonfigYamlString(loadedKonfigYaml))
  CliUx.ux.action.stop()
}
