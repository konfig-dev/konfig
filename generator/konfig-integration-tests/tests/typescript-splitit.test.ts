import { e2e } from "../util";
import { test } from "vitest";

test("splitit/typescript-splitit", async () => {
  await e2e(4022);
});
