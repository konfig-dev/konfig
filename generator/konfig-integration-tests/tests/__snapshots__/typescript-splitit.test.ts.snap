// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`splitit/typescript-splitit 1`] = `
"# splitit-web-typescript-sdk<a id=\\"splitit-web-typescript-sdk\\"></a>

Splitit\\\\'s Web API

[![npm](https://img.shields.io/badge/npm-v3.3.2-blue)](https://www.npmjs.com/package/splitit-web-typescript-sdk/v/3.3.2)

## Table of Contents<a id=\\"table-of-contents\\"></a>

<!-- toc -->

- [Installing](#installing)
  * [npm](#npm)
  * [yarn](#yarn)
- [Getting Started](#getting-started)
- [Reference](#reference)
  * [\`typescriptsplitit.installmentplan.cancel\`](#typescriptsplititinstallmentplancancel)
  * [\`typescriptsplitit.installmentplan.checkEligibility\`](#typescriptsplititinstallmentplancheckeligibility)
  * [\`typescriptsplitit.installmentplan.get\`](#typescriptsplititinstallmentplanget)
  * [\`typescriptsplitit.installmentplan.post\`](#typescriptsplititinstallmentplanpost)
  * [\`typescriptsplitit.installmentplan.post2\`](#typescriptsplititinstallmentplanpost2)
  * [\`typescriptsplitit.installmentplan.refund\`](#typescriptsplititinstallmentplanrefund)
  * [\`typescriptsplitit.installmentplan.search\`](#typescriptsplititinstallmentplansearch)
  * [\`typescriptsplitit.installmentplan.updateOrder\`](#typescriptsplititinstallmentplanupdateorder)
  * [\`typescriptsplitit.installmentplan.updateOrder2\`](#typescriptsplititinstallmentplanupdateorder2)
  * [\`typescriptsplitit.installmentplan.verifyAuthorization\`](#typescriptsplititinstallmentplanverifyauthorization)

<!-- tocstop -->

## Installing<a id=\\"installing\\"></a>

### npm<a id=\\"npm\\"></a>
\`\`\`
npm install splitit-web-typescript-sdk --save
\`\`\`

### yarn<a id=\\"yarn\\"></a>
\`\`\`
yarn add splitit-web-typescript-sdk
\`\`\`

## Getting Started<a id=\\"getting-started\\"></a>

\`\`\`typescript
import { TypescriptSplitit } from \\"splitit-web-typescript-sdk\\";

const typescriptsplitit = new TypescriptSplitit({
  // Defining the base path is optional and defaults to https://web-api-v3.production.splitit.com
  // basePath: \\"https://web-api-v3.production.splitit.com\\",
  oauthClientId: \\"CLIENT_ID\\",
  oauthClientSecret: \\"CLIENT_SECRET\\",
});

const cancelResponse = await typescriptsplitit.installmentplan.cancel({
  installmentPlanNumber: \\"installmentPlanNumber_example\\",
  xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
  xSplititTouchPoint: \\"\\",
});

console.log(cancelResponse);
\`\`\`

## Reference<a id=\\"reference\\"></a>


### \`typescriptsplitit.installmentplan.cancel\`<a id=\\"typescriptsplititinstallmentplancancel\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const cancelResponse = await typescriptsplitit.installmentplan.cancel({
  installmentPlanNumber: \\"installmentPlanNumber_example\\",
  xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
  xSplititTouchPoint: \\"\\",
});
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### installmentPlanNumber: \`string\`<a id=\\"installmentplannumber-string\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentPlanCancelResponse](./models/installment-plan-cancel-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/{installmentPlanNumber}/cancel\` \`POST\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.checkEligibility\`<a id=\\"typescriptsplititinstallmentplancheckeligibility\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const checkEligibilityResponse =
  await typescriptsplitit.installmentplan.checkEligibility({
    xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
    xSplititTouchPoint: \\"\\",
  });
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

##### PlanData: [\`PlanData\`](./models/plan-data.ts)<a id=\\"plandata-plandatamodelsplan-datats\\"></a>

##### CardDetails: [\`CardData\`](./models/card-data.ts)<a id=\\"carddetails-carddatamodelscard-datats\\"></a>

##### BillingAddress: [\`AddressData\`](./models/address-data.ts)<a id=\\"billingaddress-addressdatamodelsaddress-datats\\"></a>

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentsEligibilityResponse](./models/installments-eligibility-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/check-eligibility\` \`POST\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.get\`<a id=\\"typescriptsplititinstallmentplanget\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const getResponse = await typescriptsplitit.installmentplan.get({
  installmentPlanNumber: \\"installmentPlanNumber_example\\",
  xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
  xSplititTouchPoint: \\"\\",
});
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### installmentPlanNumber: \`string\`<a id=\\"installmentplannumber-string\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentPlanGetResponse](./models/installment-plan-get-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/{installmentPlanNumber}\` \`GET\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.post\`<a id=\\"typescriptsplititinstallmentplanpost\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const postResponse = await typescriptsplitit.installmentplan.post({
  xSplititTestMode: \\"None\\",
  xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
  xSplititTouchPoint: \\"\\",
  AutoCapture: true,
});
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### AutoCapture: \`boolean\`<a id=\\"autocapture-boolean\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

##### Attempt3dSecure: \`boolean\`<a id=\\"attempt3dsecure-boolean\\"></a>

##### Shopper: [\`ShopperData\`](./models/shopper-data.ts)<a id=\\"shopper-shopperdatamodelsshopper-datats\\"></a>

##### PlanData: [\`PlanDataModel\`](./models/plan-data-model.ts)<a id=\\"plandata-plandatamodelmodelsplan-data-modelts\\"></a>

##### BillingAddress: [\`AddressDataModel\`](./models/address-data-model.ts)<a id=\\"billingaddress-addressdatamodelmodelsaddress-data-modelts\\"></a>

##### RedirectUrls: [\`InitiateRedirectionEndpointsModel\`](./models/initiate-redirection-endpoints-model.ts)<a id=\\"redirecturls-initiateredirectionendpointsmodelmodelsinitiate-redirection-endpoints-modelts\\"></a>

##### UxSettings: [\`UxSettingsModel\`](./models/ux-settings-model.ts)<a id=\\"uxsettings-uxsettingsmodelmodelsux-settings-modelts\\"></a>

##### EventsEndpoints: [\`EventsEndpointsModel\`](./models/events-endpoints-model.ts)<a id=\\"eventsendpoints-eventsendpointsmodelmodelsevents-endpoints-modelts\\"></a>

##### ProcessingData: [\`ProcessingData\`](./models/processing-data.ts)<a id=\\"processingdata-processingdatamodelsprocessing-datats\\"></a>

##### xSplititTestMode: \`'None' | 'Regular' | 'Fast' | 'Automation'\`<a id=\\"xsplitittestmode-none--regular--fast--automation\\"></a>

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InitiatePlanResponse](./models/initiate-plan-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/initiate\` \`POST\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.post2\`<a id=\\"typescriptsplititinstallmentplanpost2\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const post2Response = await typescriptsplitit.installmentplan.post2({
  xSplititTestMode: \\"None\\",
  xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
  xSplititTouchPoint: \\"\\",
  AutoCapture: true,
  TermsAndConditionsAccepted: true,
});
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### AutoCapture: \`boolean\`<a id=\\"autocapture-boolean\\"></a>

##### TermsAndConditionsAccepted: \`boolean\`<a id=\\"termsandconditionsaccepted-boolean\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

##### Attempt3dSecure: \`boolean\`<a id=\\"attempt3dsecure-boolean\\"></a>

##### Shopper: [\`ShopperData\`](./models/shopper-data.ts)<a id=\\"shopper-shopperdatamodelsshopper-datats\\"></a>

##### PlanData: [\`PlanDataModel\`](./models/plan-data-model.ts)<a id=\\"plandata-plandatamodelmodelsplan-data-modelts\\"></a>

##### BillingAddress: [\`AddressDataModel\`](./models/address-data-model.ts)<a id=\\"billingaddress-addressdatamodelmodelsaddress-data-modelts\\"></a>

##### PaymentMethod: [\`PaymentMethodModel\`](./models/payment-method-model.ts)<a id=\\"paymentmethod-paymentmethodmodelmodelspayment-method-modelts\\"></a>

##### RedirectUrls: [\`RedirectionEndpointsModel\`](./models/redirection-endpoints-model.ts)<a id=\\"redirecturls-redirectionendpointsmodelmodelsredirection-endpoints-modelts\\"></a>

##### ProcessingData: [\`ProcessingData\`](./models/processing-data.ts)<a id=\\"processingdata-processingdatamodelsprocessing-datats\\"></a>

##### EventsEndpoints: [\`EventsEndpointsModel\`](./models/events-endpoints-model.ts)<a id=\\"eventsendpoints-eventsendpointsmodelmodelsevents-endpoints-modelts\\"></a>

##### xSplititTestMode: \`'None' | 'Regular' | 'Fast' | 'Automation'\`<a id=\\"xsplitittestmode-none--regular--fast--automation\\"></a>

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentPlanCreateResponse](./models/installment-plan-create-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans\` \`POST\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.refund\`<a id=\\"typescriptsplititinstallmentplanrefund\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const refundResponse = await typescriptsplitit.installmentplan.refund({
  installmentPlanNumber: \\"installmentPlanNumber_example\\",
  xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
  xSplititTouchPoint: \\"\\",
  Amount: 3.14,
  RefundStrategy: \\"FutureInstallmentsFirst\\",
});
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### Amount: \`number\`<a id=\\"amount-number\\"></a>

##### installmentPlanNumber: \`string\`<a id=\\"installmentplannumber-string\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

##### RefundStrategy:<a id=\\"refundstrategy\\"></a>

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentPlanRefundResponse](./models/installment-plan-refund-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/{installmentPlanNumber}/refund\` \`POST\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.search\`<a id=\\"typescriptsplititinstallmentplansearch\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const searchResponse = await typescriptsplitit.installmentplan.search({
  xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
  xSplititTouchPoint: \\"\\",
});
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

##### installmentPlanNumber: \`string\`<a id=\\"installmentplannumber-string\\"></a>

##### refOrderNumber: \`string\`<a id=\\"refordernumber-string\\"></a>

##### extendedParams: Record<string, \`string\`><a id=\\"extendedparams-record\\"></a>

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentPlanSearchResponse](./models/installment-plan-search-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/search\` \`GET\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.updateOrder\`<a id=\\"typescriptsplititinstallmentplanupdateorder\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const updateOrderResponse = await typescriptsplitit.installmentplan.updateOrder(
  {
    installmentPlanNumber: \\"installmentPlanNumber_example\\",
    xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
    xSplititTouchPoint: \\"\\",
    ShippingStatus: \\"Pending\\",
  }
);
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### installmentPlanNumber: \`string\`<a id=\\"installmentplannumber-string\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

##### TrackingNumber: \`string\`<a id=\\"trackingnumber-string\\"></a>

##### RefOrderNumber: \`string\`<a id=\\"refordernumber-string\\"></a>

##### ShippingStatus:<a id=\\"shippingstatus\\"></a>

##### Capture: \`boolean\`<a id=\\"capture-boolean\\"></a>

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentPlanUpdateResponse](./models/installment-plan-update-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/{installmentPlanNumber}/updateorder\` \`PUT\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.updateOrder2\`<a id=\\"typescriptsplititinstallmentplanupdateorder2\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const updateOrder2Response =
  await typescriptsplitit.installmentplan.updateOrder2({
    xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
    xSplititTouchPoint: \\"\\",
    ShippingStatus: \\"Shipped\\",
  });
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

##### RefOrderNumber: \`string\`<a id=\\"refordernumber-string\\"></a>

##### TrackingNumber: \`string\`<a id=\\"trackingnumber-string\\"></a>

##### Capture: \`boolean\`<a id=\\"capture-boolean\\"></a>

##### ShippingStatus:<a id=\\"shippingstatus\\"></a>

##### Identifier: [\`IdentifierContract\`](./models/identifier-contract.ts)<a id=\\"identifier-identifiercontractmodelsidentifier-contractts\\"></a>

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[InstallmentPlanUpdateResponse](./models/installment-plan-update-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/updateorder\` \`PUT\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


### \`typescriptsplitit.installmentplan.verifyAuthorization\`<a id=\\"typescriptsplititinstallmentplanverifyauthorization\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`typescript
const verifyAuthorizationResponse =
  await typescriptsplitit.installmentplan.verifyAuthorization({
    installmentPlanNumber: \\"installmentPlanNumber_example\\",
    xSplititIdempotencyKey: \\"xSplititIdempotencyKey_example\\",
    xSplititTouchPoint: \\"\\",
  });
\`\`\`

#### ⚙️ Parameters<a id=\\"⚙️-parameters\\"></a>

##### installmentPlanNumber: \`string\`<a id=\\"installmentplannumber-string\\"></a>

##### xSplititIdempotencyKey: \`string\`<a id=\\"xsplititidempotencykey-string\\"></a>

##### xSplititTouchPoint: \`string\`<a id=\\"xsplitittouchpoint-string\\"></a>

TouchPoint

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[VerifyAuthorizationResponse](./models/verify-authorization-response.ts)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/api/installmentplans/{installmentPlanNumber}/verifyauthorization\` \`GET\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


## Author<a id=\\"author\\"></a>
This TypeScript package is automatically generated by [Konfig](https://konfigthis.com)
"
`;
